---
title: "NemaScan Report for Cry5B"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: true
    toc_depth: 5
    toc_float:
      collapsed: true
      smooth_scroll: true
---

<br>

## Usage

* This R Markdown file imports data tables within Analysis_Results-Date folder and will knit without any modification.

<br>


```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE)
options(knitr.duplicate.label = 'allow')
knitr::opts_knit$set(progress = FALSE)
```


```{r include=FALSE}
library(tidyverse)
library(glue)
library(ggnewscale)
library(ggbeeswarm)
library(gganimate)

# install gganimate
# https://r-graph-gallery.com/271-ggplot2-animated-gif-chart-with-gganimate.html

# library("knitr", lib.loc = library.path)
# library("ggrepel", lib.loc = library.path)
# library("genetics", lib.loc = library.path)
# library("cowplot", lib.loc = library.path)
# library("stringr", lib.loc = library.path)

```

## Manhattan plot

A genome-wide association study (GWAS) was performed by testing whether marker genotype differences can explain phenotypic variation. These tests correct for relatedness among individuals in the population using a genomic relatedness matrix (or "kinship matrix"). This anlaysis was performed with GCTA using two different kinship matrices: one constructed specifically with inbred model organisms in mind (INBRED) and one which is constructed from all markers except those on the chromosome of the tested marker ("leave-one-chromosome-out"; LOCO). The INBRED kinship matrix more heavily corrects for genetic stratification at the tested marker, while the LOCO kinship matrix does not, and may therefore increase power in certain contexts.

* **Every dot is a SNV marker.**

* **SNVs are colored if they pass the genome-wide corrected significance threshold:**
    * The horizontal solid line corresponds to stricter Bonferroni (BF) threshold which is based on the number of markers in the analysis.
    * The horizontal dash line corresponds to more permissive EIGEN threshold, which corrects for the number of independent markers in your data set. This threshold takes advantage of the extensive LD in *C. elegans* to limit the number of "unique" markers. (See Zdraljevic *et al*. 2019 (PMID: 30958264) for more)
    * If you selected a custom threshold, only this threshold is shown as a dotted line.

```{r, fig.height=6, fig.width=12}
# load processed mapping data. 
# Note that readr::read_delim will throw parsing errors and put NA in rows that contain actual values in strain/value/allele/var.exp so did not use it
# load inbred mapping reslts
processed_mapping <- read.delim("processed_Cry5B_inbred.tsv", stringsAsFactors=FALSE) %>%
    dplyr::mutate(CHROM = factor(CHROM, levels = c("I","II","III","IV","V","X","MtDNA"))) %>%
    dplyr::select(-marker) %>%
    tidyr::unite("marker", CHROM, POS, sep = ":", remove = F) %>%
    dplyr::mutate(algorithm = "INBRED")

# do we have mito mapping?
mito_check <- processed_mapping %>%
    na.omit()

## MANHATTAN PLOTS ##
for.plot <- processed_mapping %>%
  dplyr::mutate(CHROM = as.factor(CHROM)) %>%
    {
        if(!("MtDNA" %in% mito_check$CHROM)) dplyr::filter(., CHROM != "MtDNA") else .
    }
BF <- processed_mapping %>% 
    dplyr::group_by(trait, algorithm) %>% 
    dplyr::filter(log10p != 0) %>% 
    dplyr::distinct(marker, log10p) %>%
    dplyr::mutate(BF = -log10(0.05/sum(log10p > 0, na.rm = T))) %>%
    dplyr::ungroup() %>%
    dplyr::select(BF) %>%
    unique(.) %>%
    dplyr::slice(1) %>% # BF can be slightly different between loco and inbred... but just plot one (5.46 v 5.47...)
    as.numeric()

BF.frame <- processed_mapping %>%
  dplyr::select(trait) %>%
  dplyr::filter(!duplicated(trait)) %>%
  dplyr::mutate(BF = BF, user = unique(processed_mapping$BF)[1])

# if user selected a different threshold, use that, otherwise plot BF and EIGEN
if(BF.frame$user %in% c(BF.frame$BF)) {
  for.plot.ann <- for.plot %>%
    dplyr::mutate(sig = case_when(log10p > BF.frame$BF ~ "BF",
                                  TRUE ~ "NONSIG"))
  
  sig.colors <- c("red","black")
  names(sig.colors) <- c("BF","NONSIG")
} else {
  for.plot.ann <- for.plot %>%
    dplyr::mutate(sig = case_when(log10p > BF.frame$user ~ "user",
                                  TRUE ~ "NONSIG"))
  
  sig.colors <- c("red", "black")
  names(sig.colors) <- c("user", "NONSIG")
}

test <- BF.frame %>%
    tidyr::pivot_longer(BF:user) %>%
    dplyr::distinct() %>%
    dplyr::filter(name %in% names(sig.colors))

# are we plotting mito or no?
if("MtDNA" %in% unique(for.plot.ann$CHROM)) {
    facet_scales <- "fixed"
} else {
    facet_scales <- "free"
}


# make the plot

man.plot <- ggplot() + 
  theme_bw() + 
  geom_point(data = for.plot.ann, 
             mapping = aes(x = POS/1000000, 
                           y = log10p,
                           colour = sig,
                           alpha = sig)) +
  scale_alpha_manual(values = c("BF" = 1,  "user" = 1, "NONSIG" = 0.25)) +
  scale_colour_manual(values = sig.colors) + 
  scale_x_continuous(expand = c(0, 0), breaks = c(5, 10, 15, 20)) +
  geom_hline(data = test, aes(yintercept = value, linetype = name)) + 
  scale_linetype_manual(values = c("BF" = 1, "user" = 2)) +
  labs(x = "Genomic position (Mb)",
       y = expression(-log[10](italic(p)))) +
  theme(legend.position = "none", panel.grid = element_blank()) + 
  facet_grid(algorithm ~ CHROM, scales = "free_x", space = facet_scales) + 
  ggtitle(BF.frame$trait)


# use code below to save the plot to file. 
# ggsave(man.plot, filename = paste0(BF.frame$trait,"_manhattan.plot.png"), width = 8, height = 4)


# show plot 
man.plot
```

### Phenotype by Genotype Split

For each detected QTL, we can observe the phenotypes of the strains with the reference (REF) allele (i.e. same genotype as N2) compared to the phenotypes of the strains with the alternative (ALT) allele (i.e. genotype different than N2). A QTL is defined as a region where genetic variation is correlated with phenotypic variation, so we expect to see a difference in phenotype between the REF and ALT groups. In a best-case scenario, we like to see a large split between REF and ALT and a good number of strains in both groups. It is also important to ensure that the mean phenotype of neither group is driven by a small number of outlier strains. 

*A few select strains are highlighted due to their use in Andersen Lab dose response assays*

```{r, eval=whether_plot_pxg}
nested.pxg.dat <- processed_mapping %>%
    dplyr::filter(algorithm == "INBRED") %>%
    dplyr::filter(!is.na(peak_id)) %>%
    dplyr::select(CHROM, marker, trait, startPOS, peakPOS, endPOS, AF1, value, strain, allele, peak_id) %>%
    dplyr::distinct() %>%
    dplyr::mutate(startPOS = startPOS/1000000,
                  peakPOS = peakPOS/1000000,
                  endPOS = endPOS/1000000) %>%
    # dplyr::left_join(.,sweep.chrom.pivot) %>% # don't have this file yet
    # dplyr::group_by(trait, peak_id) %>%
    # dplyr::recode(allele, "-1" = "REF", "1" = "ALT") %>%
    dplyr::mutate(allele = dplyr::case_when(allele == "-1" ~ "REF",
                                            allele == "1" ~ "ALT",
                                            TRUE ~ "NA"),
                  allele = factor(allele, levels = c("REF", "ALT")))

strains.of.interest <- c("PD1074", "N2", "CB4856", "RC301", "MY16", 
                             "ECA396", "ECA36", "XZ1516", "ECA248", "AB1", 
                             "CB4507", "CB4858", "CB4855", "CB4852", "MY1", 
                             "JU319", "JU345", "JU400", "PB306", "PX174", "PX179")


plot <- nested.pxg.dat %>%
    dplyr::filter(allele != "NA" | !is.na(allele)) %>%
    dplyr::mutate(SOI = strain %in% strains.of.interest,
                  SOI.3 = dplyr::case_when(strain %in% c("N2", "PD1074") ~ "N2",
                                           strain == "CB4856" ~ "CB",
                                           strain %in% strains.of.interest ~ "special",
                                           TRUE ~ "other"),
                  SOI.2 = if_else(SOI == TRUE, true = strain, false = "")) %>%
    droplevels() %>%
    dplyr::arrange(SOI.2) %>%
    ggplot2::ggplot(mapping = aes(x = allele, y = value, text = SOI.2)) +
    ggplot2::theme_bw(base_size = 12) +
    ggplot2::geom_violin(aes(fill = allele), alpha = 0.5, scale = "count", draw_quantiles = c(0.25, 0.5, 0.75)) +
    ggplot2::scale_fill_manual(values = c("REF" = "#A79F92", "ALT" = "mediumpurple4"), guide = FALSE) +
    ggnewscale::new_scale("fill") +
    ggplot2::geom_point(aes(fill = SOI.3, size = SOI), position = ggbeeswarm::position_beeswarm(), shape = 21) +
    ggplot2::scale_fill_manual(values = c("N2" = "orange", "CB" = "blue", "special" ="red", "other" ="grey50"), guide = FALSE) +
    ggplot2::scale_size_manual(values = c(1.5,2.5), guide = FALSE) +
    ggrepel::geom_text_repel(aes(label = SOI.2),
                             colour = "black", position = ggbeeswarm::position_beeswarm()) +
    ggplot2::theme(legend.position = "bottom") +
    ggplot2::labs(y = "Trait Value",
                  x = "Genotype") +
    ggplot2::facet_grid(~marker)

print(plot)

# save
# ggsave(glue::glue("pxg.png"), height = 5, width = 5)
```


